### DISCUSSION TITLE
A modal for asking values from user (Prompt)

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#17] [Ideas]
Like I mentioned in [a comment in issue #14](https://github.com/Taitava/obsidian-shellcommands/issues/14#issuecomment-920159089), I have been thinking about a new variable named ~~{{input}}~~. When a command is executed, if the variable is present in the command, the plugin could open an small popup window and ask the user to type something, that will then be the value of ~~{{input}}~~.

I was originally thinking only about a typeable freeform value, but @FelipeRearden mentioned that ~~{{input}}~~ could also have a list of options that the user can select from. Really good, as sometimes you do need to limit the possibilities so that it's easy for the user to decide what to input. So I started to think that there needs to be a way to express what kind of form element the user is offered.

## Added 2022-02-07: No {{input}} variable, but custom variables instead

This post talks a lot about a variable named `{{input}}`. I've decided not to implement that variable. The inputted values will be stored to [custom variables](https://github.com/Taitava/obsidian-shellcommands/discussions/146), whose names can be decided by a user. This way the inputted values will have more meaningful variables names in the shell command definitions. E.g. if you ask for a due date, you can name the variable to e.g. `{{_due_date}}`, and then a shell command's source code `echo "The due date is:" {{_due_date}}` tells clearly that the variable contains a date. `echo "The due date is:" {{input}}` would not be that descriptive.

-----

- A bullet point item: Planned
- [ ] An empty box: Will be implemented next.
- [x] A checked box: Implemented, but not yet released.
- ✔️: Implemented and released.

## Input types
- ✔️ A single line text field.
- ✔️ A password field. Thank you @SrGeneroso for the idea!
- ✔️ A multiline text field. (~~How well do linebreak characters work in shell commands? That I'll have to see. In some cases they might end one command and start a new one, unless they are inside quotes~~). * Edit 2022-05-12: Linebreaks will work just fine, the current method to escape special characters - including linebreaks - handles this well.
- ✔️ A dropdown list of predefined options, like @FelipeRearden suggested. Allows one item to be selected.
- A multi-option list?? Allows multiple items to be selected. Maybe it would be troublesome to implement (what to use as a separator between multiple selected options), so I think for the beginning a freeform text field is a better alternative for a multi-option list. *
- ✔️ On/off toggles (aka checkboxes). These need to have a definition for two texts: a text when the toggle is on, and a text when the toggle is off. *
- File selector. With the possibility to limit the selection to files contained in a specific folder (how about files in that folder's subfolders?).
- Folder selector. With the possibility to limit the selection to folders contained in a specific folder (how about folders in that folder's subfolders?).
- Color field, maybe? Does anyone need it? Please comment down below if you would need it.
- Anything else?

*) Would not be implemented in the first version of ~~{{input}}~~.

## Definitions
I'd like the input elements to have certain properties:
- ✔️ Description: The user should be displayed a description text so that they know what to input and how their decision affects the outcome. If description is empty, the user will be shown the actual shell command (maybe not an alias, even if an alias is defined) and the position where their input is going to be placed.
- ✔️ Input type. E.g. `singleline`/`multiline`/`list`/`toggle`.
- ✔️ Default value. For lists, this would be the default selected option.
- ✔️ Ability to have field type specific properties.

~~The following is just a draft example of the definition format, but I'd just like to give you a clue what it might look like. It will probably change.~~
~~`{{input:singleline:Description text goes here:This is the default value}}`~~

## Multiple ~~{{input}}~~'s in one command
If a command has multiple ~~{{input}}~~ variables in it, all of them will be displayed in the same modal (= popup window). No need to open a new modal one after each other. It's easier for the user to see all the inputs at once, and if they have filled one field, then switch to another, they might change their mind regarding the first field, and it's easy to go back to edit values they already filled.

For example:
~~`cd {{input:singleline:Give a directory for the new file}} && touch {{input:singleline:Give a name for the new file}}`~~
Edit 2022-03-24: An updated example:
`cd {{_directory_name}} && touch {{_new_file_name}}`
Prompts will be defined separately in the settings, so the questions will not need to be declared in the {{variables}}. Custom variables will be used to pass the values, which means that the names `{{_directory_name}}` and `{{_new_file_name}}` are freely decideable.

## Ability to refer to the already asked ~~{{input}}~~ a second time
~~For example:~~
~~`mkdir {{directory=input:singleline:Give a directory for the new file}} cd {{custom:directory}} && touch {{input:singleline:Give a name for the new file}}`~~

~~Here `{{directory=input...` would define a custom variable for us and store the result of {{input}} there so that we can use the same value twice without asking the user twice. `{{custom:directory}}` would refer to the `directory` variable we just created. I could also consider accessing custom variables without the `custom:` prefix, simply like `{{my_custom_variable}}`, but I fear that there might be overlapping with variable names if I introduce new variables for this plugin in the future. Let's go with `{{custom:my_variable_name}}` first and decide later if the shorter format is needed.~~

~~But talking about *first*... this ability to use custom variables would be separate feature issue, and it would be implemented later than the first version of {{input}}~~.

Edit 2022-03-24: The plan in this chapter is outdated. But anyway, there will be a way to use an inputted value multiple times in a shell command. Just use a custom variable multiple times, e.g. `touch {{_new_file_name}} && gedit {{_new_file_name}}`.

**Comment by Taitava** [comment]
I think another way to implement would be to make this a *preaction* instead of a *variable*, mainly for two reasons:
 - As a variable, this would be more than just reading values of certain properties. This would be about showing a user a modal asking for input, which pauses the workflow of variable parsing, and I would need to design how to implement this kind of pauses in the varible parsing process. I think this would be so clean design as a variable.
 - A preaction might offer better circumstances for customising an input field. There could be a real settings form for defining the input element's attributes (type: singleline/multiline/list/toggle, name, description, default value, max length, required to fill or can be left empty? etc.). Those could be quite annoying to define in a text format `{{input:singleline:Give a directory for the new file:plus:other:arguments}}`.

I have not yet created a discussion for *preactions*, but will do it sometime.

**Reply by FelipeRearden** [reply]
After reading this, from the user standpoint, preactions seems to be the best path. 

I think the most important feature for `{{input}}` is the ability to customize the input element's attributes for each need (each SC workflow)

**Comment by FelipeRearden** [comment]
The content of this post was moved to another discussion in here #74 !

**Reply by Taitava** [reply]
I think this is actually a different idea than what the `{{input}}` idea is, so it could be discussed in a new discussion. `{{input}}` would come into play after a user has selected a shell command that should be executed. Your idea is about selecting which shell command to run. Anyway, it's a good idea. Can you create a new discussion for it, please? Thanks! 🙂

**Reply by FelipeRearden** [reply]
Thanks @Taitava !!! Now I understood correctly what an input is :)

Done in  #74  :)

Have a great day!

**Comment by FelipeRearden** [comment]
Hello @Taitava !!!!

Let's see if this time I have a user case for {{input}} :)

## User Case - Append text from Active Note to one Note from a List of Notes

01 I have a note that is an inbox of all my twitter bookmarks:
- I created a Shortcut on iOS to send a tweet that I chose to an specific note in Obsidian 
- the tweet is imported with a standard formation. Every tweet have the same structure in the note

02 When I'm reading in this note, I want to transfer some content for a note related to the tweet. 

03 Example I have a tweet from Ryan Holiday that is a quote from Marcus Aurelius that I want to send to the note [[Marcus Aurelius]]
![708D3C51-05D7-4E99-A578-2D9B258C2606](https://user-images.githubusercontent.com/85180696/138442420-ec7c5cbd-2bee-45d2-af25-b5b4fbc247d2.jpeg)


I have a SC to do this: `echo "\n{{selection}} >> "Marcus Aurelius.md"`

The same spots described above is applied to 5 notes for example:
- [[Marcus Aurelius]]
- [[Seneca]]
- [[Nassim Taleb]]
- [[Epitectus]]
- [[Mark Manson]]

My idea is to use the {{input}} to select the destination folder of each tweet:

`echo "\n{{selection}} >> "{{input}}.md"`

**Options for {{input}}**

- Option 1: Where {{input}} is a list of these 5 notes previously defined (best scenario for this specific case)
- Option 2: A modal where the user type/input the destination note. 


Is this the right example of {{input}} ?

**Reply by Taitava** [reply]
I think it is a good example. 👍



>     * Option 1: Where {{input}} is a list of these 5 notes previously defined (best scenario for this specific case)

I think the previously defined _A dropdown list of predefined options_ could be used for this?

>     * Option 2: A modal where the user type/input the destination note.

I added 
> - File selector.
> - Folder selector.

to the list of _Input types_, based on your idea.

Did I understood correctly what you need? 🙂 

P.S. My weekend will be busy and I might not have time to read or answer discussions before next week (I have still couple of your messages unread atm). But I'll get back to you! 😎

**Reply by FelipeRearden** [reply]
> I think it is a good example. 👍

Great !!!!!!!

> I think the previously defined _A dropdown list of predefined options_ could be used for this?

Yes! It's exactly a dropdown list !!!!!  

> I added 
> - File selector.
> - Folder selector. 
> to the list of _Input types_, based on your idea.

Oowww. Great idea @Taitava!

> Did I understood correctly what you need? 🙂 

Yes @Taitava ! it's perfect !
 
> P.S. My weekend will be busy and I might not have time to read or answer discussions before next week (I have still couple of your messages unread atm). But I'll get back to you! 😎 

Thanks for letting me know @Taitava . I really wish you a fantastic weekend I hope you get all success possible in this busy weekend :)

We talk soon :)

**Reply by Taitava** [reply]
Thanks and have a nice week! 🙂

**Comment by omnosti** [comment]
I would love this feature for being able to change the last-modified date for a note without leaving Obsidian. I think I'd be able to use this to run a command like `$(Get-Item "{{file_name}}").lastwritetime=$(Get-Date "{{input}}")`.

**Reply by Taitava** [reply]
Really good example! 🙂

Now that I have been thinking about this feature for some months, I think I'm starting to be quite ready for planning the implementation for this maybe for version `0.12.0` ~~`0.11.0` (so, after #123 is done and `0.10.0` is released)~~. There will be some changes to how the `{{input}}` variable will actually be used, compared to the early plans above, but the overall features will be available.

Edit 2022-02-22: Changed the milestone to `0.12.0`.

**Reply by Taitava** [reply]
# Still in progress, changing the milestone

**I will actually move this to the `0.12.0` milestone**, as there have been (and still is) quite a lot of other, related development that this feature is dependent on:
- #146  
- #160 

Just a sneak peek of the development so far. Here's a simple prompt. Even though the promp is already visible, the values are not yet stored into any variables - there's quite a lot of work still left to make that happen.
![kuva](https://user-images.githubusercontent.com/13002142/155079348-86320563-b307-4ec5-bbcf-22402e5c2ee1.png)
- All those three fields are configured in the settings UI, title too.
- All the values shown are default values configured in settings, too.
- In the first phase, single line text fields are supported. More field types will come some time after the first release of this feature.
- Fields can be optional or required. In the next image I cleared the first field, tried to click the _Execute_ button, and got an error message:
![kuva](https://user-images.githubusercontent.com/13002142/155080162-704c557d-1113-4a6b-95f0-8223f62ae8f5.png)

**Reply by Taitava** [reply]
# Progress, progress...

I just wanted to post something as this is still taking some time to be finished.

Accomplished features:
- Values are now stored to custom variables and shell commands can read them.
- An ability to preview prompts when configuring them (= no need to execute a shell command to get a prompt open).
- Prompt title, prompt description, field names and field descriptions can all use `{{variables}}`.
- The prompt can show a preview of the shell command with variable values inserted from the fields. The preview is updated when typing, and the currently focused field's value is **bolded** in the preview.
- Text fields are now enlarged to 50% of the modal's width.
- Modify the layout of your prompts with custom CSS: Each prompt modal has their own, unique CSS class.
- _Execute_ button text can be changed to whatever you want.

![kuva](https://user-images.githubusercontent.com/13002142/159689674-94ff80a2-3b7f-4bc5-a84e-a7facce47d36.png)

**Reply by Taitava** [reply]
# Beta test
@omnosti this feature is now included in a beta test: https://github.com/Taitava/obsidian-shellcommands/discussions/201

**Comment by Taitava** [comment]
# I renamed the discussion
Old title: _An {{input}} variable for asking a value from user_
New title: _A modal for asking values from user_

**Comment by OMSmolina** [comment]
I don't understand the description of the Folder selector with the limits but I would like to be able to make a prompt that can select any folder on the system and send that path to the script or command.💪🏻

**Reply by Taitava** [reply]
For example, if you would have the following folder structure in your Obsidian vault:

```
Vault's root folder
  - Books
    - Fiction
    - Non-fiction
  - Movies
    - Fiction
    - Documentaries
```

You could then (optionally) define that only a folder related to books can be selected, so the field would only allow selecting one of:
- `Books`
- `Books/Fiction`
- `Books/Non-fiction`

> I would like to be able to make a prompt that can select any folder on the system

Ok, so you would like to be able to select a folder that can be outside of the Obsidian vault? It's a good point, and I could make it possible.

There could also be a setting that could allow selecting non-existing folders. I.e. it would allow typing a name for a new folder that will be created by the executed script.

**Reply by OMSmolina** [reply]
Ah, now I understand, thank you!

Yes, I would like to be able to choose the path of a folder on the system.

For example to send that path to a script `script.bat /path/selected/in/propmt`

Another example could be a symlink to a folder `mklink /path/selected/in/propmt  /path2/selected2/in2/propmt2`

In principle, both the field to search for a folder or file should be quite easy and fast since you end up using the system file explorer when opening a file/folder. The limits and advanced things that you have explained to me would take a little more time.

**Comment by Taitava** [comment]
# New field types are now released in `0.21.0`
- Password field
- Multiline text field
- Dropdown list
- On/off toggle field

Previously, only single line text fields were available.
